Fix Circular References

b-butler opened the pull request:
## Description

This fixes memory leaks associated with using a logger and perhaps other parts of HOOMD's interface using `weakref`. Why this is necessary, I don't know. I think it may have something to do with a C++ object holding a reference to a Python object which holds a reference to the C++ object, but that is purely conjecture.

This moves simulation references to weak references generally and moves the logger to use weak references for objects expected to live beyond the call. Unfortunately simply always wrapping leads to cases like
```python
logger[("new", "loggable")] = (lambda : 42, "scalar")
```
not working since the only reference to the function is in the logger. Similarly and perhaps more damningly,

```python
def add_quantities(logger):
    log_quantities = LogValues()
    logger["A"] = (log_quantities, "A")
    logger["B"] = (log_quantities, "B")
    logger["C"] = (log_quantities, "C")
```

would not work
<!-- Describe your changes in detail. -->

## Motivation and context

This is to fix multiple issues we have had with the garbage collector such as #1457.
<!--- Why is this change required? What problem does it solve? -->

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #1457

## How has this been tested?

Existing tests pass which ensures the same behavior as before (well tests will pass when I make this not a draft).
<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Fixed: Potential memory leaks.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

b-butler committed:
refactor: Use weakreference to simulation.

b-butler committed:
refactor: Store simulation in synced list as weakref

b-butler committed:
fix: Error when default properties are not callable

This doesn't error but is bad practice.

b-butler committed:
refactor: Use ReferenceType in logging _LoggerEntry

This removes potentially circular (viscously so given the Python memory
leak).

b-butler committed:
fix: Remove deepcopy in _SafeNamespaceDict

This fixes an error where we deepcopied and thus disconnected
_HOOMDBaseObject logggables. This also should increase the performance
of the logger.

b-butler committed:
test: Fix check function with default HOOMD obj attrs

b-butler committed:
fix: Rename SyncedList __simulation to _simulation_

Avoid python name mangling

b-butler committed:
test: Fix tests for new weak references

b-butler committed:
fix: Switch _HOOMDBaseObject to _simulation_ over __simulation

b-butler committed:
fix: Logger with weak references

1. Switch to only use weak references in HOOMD or subclassed from HOOMD
   classes.
2. Add an _InvalidLogEntry sentinel to indicate when an object has been
   deleted which the Logger pointed to.
3. Prune dead connections in the logger.

b-butler committed:
test: Fix logger test imports

tommy-waltmann commented on hoomd/data/syncedlist.py:
```diff
@@ -212,6 +213,20 @@ def _validate_or_error(self, value):
+        sim = self._simulation_
+        if sim is not None:
+            sim = sim()
+            if sim is not None:
+                return sim
```
What should this property return in the case that `sim is None` after dereferencing the weakref to `sim`? I can't imagine code that uses the `_simulation` property will want to always have to double check to make sure `_simulation` actually returned something.

joaander replied:
As far as I understand it, that case should never happen. All objects holding this are held by a `Simulation`. `_simulation` is a back reference to that object. When `_simulation` returns None (the implied return value when there is no `return` statement), it is the same as if the object has not yet been added to a Simulation.

We could add this check for completeness. The only way to trigger it would be to directly access internal data structures and bypass HOOMD's attach/detach and/or add/remove APIs.
```python
            if sim is not None:
                return sim
            else:
                raise RuntimeError("Parent Simulation was garbage collected.")
```

b-butler replied:
Yeah, we could error here or make sure we are detached (which would be my preference. This could happen in some unusual cases where the operation's lifespan is longer than the simulation's.

tommy-waltmann replied:
In my original comment, I didn't think about the implied return value of `None` when there is no return statement, but I like the idea of fully checking for attachment when the weakref points to `None`

tommy-waltmann commented on hoomd/operation.py:
```diff
@@ -310,6 +312,21 @@ def _attach(self, simulation):
+        sim = self._simulation_
+        if sim is not None:
+            sim = sim()  # grab weakref
+            if sim is not None:
+                return sim
```
Same as above comment.

tommy-waltmann reviewed (changes requested):
Have the changes in this PR been tested to make sure the memory leaks are removed?

joaander commented:
> Have the changes in this PR been tested to make sure the memory leaks are removed?

Yes. It solves the issue in both `hoomd-validation` and the case in #1457.

b-butler committed:
fix: Object behavior when outliving simulation

Now explicitly detach.

b-butler committed:
test: The new simulation weak reference and detaching

b-butler committed:
fix: Remove final standard references to simulation.

All references to simulation in all objects in HOOMD itself should now
be weakreference and not involve circular references.

b-butler committed:
fix: Ensure SyncedList unsyncs when sim is collected.

b-butler committed:
fix: Set SyncedList._simulation_ on initialization
